### PR TITLE
Optimize bash/zsh script runtimes

### PR DESCRIPTION
**Pull request by JamyGolden**
This PR contains commits mostly from https://github.com/chriskempson/base16-shell/pull/188. I cherry-picked the commits from that PR and added 2 commits on top of that.

@j-c-m and @JStation could you sort of review this? (It's a bit of a mix of us reviewing each other). I've added two commits on top of what you did. My commits include:

- Fixing and optimising `profile_helper.fish`. It still contained references to `realpath`
- Fixing and using the same script for `base16-shell.plugin.zsh` and `base16-shell.plugin.bash`
- Adding instructions for using `base16-shell.plugin.zsh` as an oh my zsh plugin. I touched the file and saw the readme instructions were missing. I don't really know how bash_it works, so I didn't add readme instructions for that.
- @JStation I tried using XDG spec paths but I couldn't get it to work. It seemed like any XDG path was an empty string during the scripts runtime. Not exactly sure why, but I changed `~` to `$HOME` for now and put the paths in variables to make it easy to change in future.
- I introduced the concept of `$HOME/.config/base16-project` as the config path for base16-* repos. That way we don't litter the `$HOME` directory unnecessarily.

I have some questions about your PR:

---

Why did you add this? It isn't used anywhere or mentioned in the readme, was it for testing?

```bash
alias reset="command reset && [ -f ~/.base16_theme ] && .~/.base16_theme"
```

---

What exactly is this for? I understand the concept but could you elaborate on the use case? Could you add something about this that I can put in the readme?

```bash
if [ -n "$BASE16_DEFAULT_THEME" ] && [ ! -f ~/.base16_theme ]; then
  ln -s "$BASE16_SHELL/scripts/base16-$BASE16_DEFAULT_THEME.sh" \
    "~/.base16_theme"
```

---

You didn't mention anything about hooks in your PR or in the readme, could you explain a bit why (maybe in such a way that I can add it to the readme) it's necessary and how it's useful?

```
  if [ -n ${BASE16_SHELL_HOOKS:+s} ] && [ -d "${BASE16_SHELL_HOOKS}" ]; then
    for hook in $BASE16_SHELL_HOOKS/*; do
      [ -f "$hook" ] && [ -x "$hook" ] && "$hook"
    done
  fi
```

### CODE REVIEW

**Comment by j-c-m** [comment]
This is really hard to follow I think 

> This PR contains commits mostly from [chriskempson/base16-shell#188](https://github.com/chriskempson/base16-shell/pull/188). I cherry-picked the commits from that PR and added some commits of my own on top of that.
> 
> @j-c-m and @JStation could you sort of review this? (It's a bit of a mix of us reviewing each other). I've added two commits on top of what you did. My commits include:
> 
> * Fixing and optimising `profile_helper.fish`. It still contained references to `realpath`
> * Fixing and using the same script for `base16-shell.plugin.zsh` and `base16-shell.plugin.bash`
> * Adding instructions for using `base16-shell.plugin.zsh` as an oh my zsh plugin. I touched the file and saw the readme instructions were missing. I don't really know how bash_it works, so I didn't add readme instructions for that.
> * @JStation I tried using XDG spec paths but I couldn't get it to work. It seemed like any XDG path was an empty string during the scripts runtime. Not exactly sure why, but I changed `~` to `$HOME` for now and put the paths in variables to make it easy to change in future.
> * I introduced the concept of `$HOME/.config/base16-project` as the config path for base16-* repos. That way we don't litter the `$HOME` directory unnecessarily.
> 
> I have some questions about your PR:
> 
> Why did you add this? It isn't used anywhere or mentioned in the readme, was it for testing?
> 
> ```shell
> alias reset="command reset && [ -f ~/.base16_theme ] && .~/.base16_theme"
> ```

No, I added it so the theme was reapplied after a terminal reset.

> 
> What exactly is this for? I understand the concept but could you elaborate on the use case? Could you add something about this that I can put in the readme?
> 
> ```shell
> if [ -n "$BASE16_DEFAULT_THEME" ] && [ ! -f ~/.base16_theme ]; then
>   ln -s "$BASE16_SHELL/scripts/base16-$BASE16_DEFAULT_THEME.sh" \
>     "~/.base16_theme"
> ```

This is how I use it...

https://github.com/j-c-m/dotfiles/blob/master/.sh/05-base16.sh

So on any new systems where I deploy my dotfiles to it will set a default based on environment.

> 
> You didn't mention anything about hooks in your PR or in the readme, could you explain a bit why (maybe in such a way that I can add it to the readme) it's necessary and how it's useful?
> 
> ```
>   if [ -n ${BASE16_SHELL_HOOKS:+s} ] && [ -d "${BASE16_SHELL_HOOKS}" ]; then
>     for hook in $BASE16_SHELL_HOOKS/*; do
>       [ -f "$hook" ] && [ -x "$hook" ] && "$hook"
>     done
>   fi
> ```

The hooks stuff was previously existing, I only removed the .vimrc_background preceding it.

**Comment by JamyGolden** [comment]
Thanks for going over it. I should have added the comments to the code lines itself to remove confusion. Creating a PR which is actually mostly someone else's PR and having them check the PR is confusing in itself though haha. 

I've simplified the changes a lot (I'll put the other changes in a different PR), what do you think about the commit I have on top of yours? I've removed `alias reset=...` since that can be added by people themselves and put into their shellrc file.

**Comment by j-c-m** [comment]
I would leave the reset alias in, I believe it is the desired and expected reset behavior when a theme is in use.
(1. call the bult-in terminal reset command 2. re-apply theme)

To make it easier to review future changes I would merge in only my changes + the single readme change around the DEFAULT variable. Follow on with seperate PR or commits for additional changes (neovim, and helper changes etc). I already noticed more dirname and readlink's sneaking back in.

File: README.md [hunk 1]
```diff
@@ -47,18 +45,20 @@ if status --is-interactive
 end
 ```
 
-Open a new shell and type `base16` followed by a tab to perform tab completion.
+## Configuration
```
Comment (JamyGolden) [inline README.md:57]: Adjust readme because `BASE16_DEFAULT_THEME` doesn't have a place under the current structure.

File: base16-shell.plugin.bash [hunk 2]
```diff
@@ -1,2 +1,2 @@
 BASE16_SHELL=$(dirname ${BASH_SOURCE[0]})
-[ -n "$PS1" ] && [ -s $BASE16_SHELL/profile_helper.sh ] && eval "$($BASE16_SHELL/profile_helper.sh)"
```
Comment (JamyGolden) [inline base16-shell.plugin.bash:2]: This would break since it's not a string anymore.

File: profile_helper.fish [hunk 3]
```diff
@@ -11,15 +11,19 @@
 # TODO: maybe port to $HOME/.config/fish/functions ?
 
 
-set SCRIPT_DIR (realpath (dirname (status -f)))
```
Comment (JamyGolden) [inline profile_helper.fish:14]: Remove dep on `realpath`

Comment (JamyGolden) [inline profile_helper.fish:14]: Remove dep on `realpath` here and another instance in this file.

File: profile_helper.fish [hunk 4]
```diff
@@ -38,3 +42,5 @@ for SCRIPT in $SCRIPT_DIR/scripts/*.sh
     end
   end
 end
+
+alias reset "command reset && [ -f ~/.base16_theme ] && sh ~/.base16_theme"
```
Comment (JamyGolden) [inline profile_helper.fish:47]: Add alias to create the same functionality for fish and bash/zsh.

File: profile_helper.fish [hunk 5]
```diff
@@ -11,15 +11,19 @@
...
 
 # load currently active theme...
 if test -e ~/.base16_theme
-  set -l SCRIPT_NAME (basename (realpath ~/.base16_theme) .sh)
-  set -gx BASE16_THEME (string match 'base16-*' $SCRIPT_NAME  | string sub -s (string length 'base16-*'))
-  eval sh '"'(realpath ~/.base16_theme)'"'
+  eval sh ~/.base16_theme
 end
 
+if test -n "$BASE16_DEFAULT_THEME"; and not test -e ~/.base16_theme
```
Comment (JamyGolden) [inline profile_helper.fish:23]: Add `$BASE16_DEFAULT_THEME` to create the same functionality for fish and bash/zsh.

File: profile_helper.sh [hunk 6]
```diff
@@ -23,12 +24,12 @@ _base16()
   fi
 }
 
-if [ -n "$BASE16_DEFAULT_THEME" ] && [ ! -f ~/.base16_theme ]; then
+if [ -n "$BASE16_DEFAULT_THEME" ] && [ ! -e ~/.base16_theme ]; then
   ln -s "$BASE16_SHELL/scripts/base16-$BASE16_DEFAULT_THEME.sh" \
-    "~/.base16_theme"
```
Comment (JamyGolden) [inline profile_helper.sh]: `~` isn't expanded in quotes.

File: profile_helper.sh [hunk 7]
```diff
@@ -23,12 +24,12 @@ _base16()
...
+    ~/.base16_theme
 fi
 
-if [ -f ~/.base16_theme ]; then
```
Comment (JamyGolden) [inline profile_helper.sh]: Check if file exists, not that it's a regular file since this is a symlink.

File: profile_helper.sh [hunk 8]
```diff
@@ -40,4 +41,4 @@ for script in "$BASE16_SHELL"/scripts/base16*.sh; do
   alias $func_name="_base16 \"${script}\" ${theme}"
 done;
 
-alias reset="command reset && [ -f ~/.base16_theme ] && .~/.base16_theme"
```
Comment (JamyGolden) [inline profile_helper.sh]: Fix spacing issue.

File: profile_helper.sh (hunk 7 above)
Comment (JamyGolden) [inline profile_helper.sh]: Check if file exists, not that it's a regular file since this is a symlink.

File: profile_helper.sh [hunk 9]
```diff
@@ -23,12 +24,12 @@ _base16()
   fi
 }
 
-if [ -n "$BASE16_DEFAULT_THEME" ] && [ ! -f ~/.base16_theme ]; then
```
Comment (JamyGolden) [inline profile_helper.sh]: Check if file exists, not that it's a regular file since this is a symlink.

**Comment by JamyGolden** [comment]
@j-c-m ok I've stripped it down to the bare essentials here. Have a look at https://github.com/base16-project/base16-shell/pull/8/commits/4d519c96aeb073ccb71659bcf03d294c75490c0d to see a comparison. I've added what I would either ask you to add/fix in your PR.
